builtin/nomad/jobspec/releaser.go

```diff
@@ -0,0 +1,547 @@
+	return r.Destroy
+}
+
+// StatusFunc implements component.Status
+func (r *Releaser) StatusFunc() interface{} {
+	return r.Status
+}
+
+func (r *Releaser) resourceManager(log hclog.Logger, dcr *component.DeclaredResourcesResp) *resource.Manager {
+	return resource.NewManager(
+		resource.WithLogger(log.Named("resource_manager")),
+		resource.WithValueProvider(r.getNomadClient),
+		resource.WithDeclaredResourcesResp(dcr),
+		resource.WithResource(resource.NewResource(
+			resource.WithName(rmResourceJobName),
+			resource.WithState(&Resource_Job{}),
+			resource.WithCreate(r.resourceJobCreate),
+			resource.WithDestroy(r.resourceJobDestroy),
+			resource.WithStatus(r.resourceJobStatus),
+			resource.WithPlatform("nomad"),
+			resource.WithCategoryDisplayHint(sdk.ResourceCategoryDisplayHint_INSTANCE_MANAGER),
+		)),
+	)
+}
+
+// getNomadClient provides
+// the client connection used by resources to interact with Nomad.
+func (r *Releaser) getNomadClient() (*nomadClient, error) {
+	// Get our client
+	client, err := api.NewClient(api.DefaultConfig())
+	if err != nil {
+		return nil, err
+	}
+	return &nomadClient{
+		NomadClient: client,
+	}, nil
+}
+
+func (r *Releaser) resourceJobCreate(
+	ctx context.Context,
+	log hclog.Logger,
+	target *Deployment,
+	result *Release,
+	state *Resource_Job,
+	client *nomadClient,
+	st terminal.Status,
+) error {
+	// Set up clients
+	jobClient := client.NomadClient.Jobs()
+	deploymentClient := client.NomadClient.Deployments()
+
+	st.Update("Getting job...")
+	jobs, _, err := jobClient.PrefixList(target.Name)
+	if err != nil {
+		return status.Errorf(codes.Aborted, "Unable to fetch Nomad jobs: %s", err.Error())
+	} else if target.Name != jobs[0].ID {
+		return status.Errorf(codes.Aborted, "Job not found: %s", err.Error())
+	}
+
+	q := &api.QueryOptions{
+		Namespace: jobs[0].JobSummary.Namespace,
+	}
+
+	job, _, err := jobClient.Info(jobs[0].ID, q)
+
+	if err != nil {
+		return status.Errorf(codes.Aborted, "Unable to fetch Nomad job: %s", err.Error())
+	}
+
+	st.Update("Getting latest deployments for job")
+	deploy, _, err := jobClient.LatestDeployment(*job.ID, q)
+	if err != nil {
+		return status.Errorf(codes.Aborted, "Unable to fetch latest deployment for Nomad job: %s", err.Error())
+	} else if deploy == nil {
+		st.Update("No active deployment for Nomad job")
+		return err
+	}
+
+	canaryDeployment := false
+	groupsToPromote := make([]string, len(deploy.TaskGroups))
+	for taskGroupName, taskGroup := range deploy.TaskGroups {
+		if r.config.Groups != nil {
+			if isElementExist(r.config.Groups, taskGroupName) && taskGroup.DesiredCanaries != 0 {
+				canaryDeployment = true
+				groupsToPromote = append(groupsToPromote, taskGroupName)
+				continue
+			}
+		} else if taskGroup.DesiredCanaries != 0 {
+			canaryDeployment = true
+			// if no groups to promote are specified in the config, promote all groups
+			// that have canaries
+			groupsToPromote = append(groupsToPromote, taskGroupName)
+		}
```

Comment:
Is this check supposed to be here, or just L123-L127 (I'm not sure what it's checking that wasn't already checked above)? Does this have the potential of a `nil` reference since we aren't checking for the existence of the task group here?

Reply:
@krantzinator this check is here in case the user did not provide any task groups in the releaser's config.

Before implementing the ability to specify particular task groups to be promoted in the releaser config, I was using the `PromoteAll` Nomad API method. So when hitting this else block, to avoid adding a task group which has no canary allocs (`DesiredCanaries == 0`) to the `groupsToPromote` slice, we verify that it has canaries. If not, we continue to the next iteration, and if none of the task groups have canaries, we exit early (L135-L138).

So in effect, the logic of this for loop & hitting this else block is "if there are no groups specified in the releaser, promote every group that has canaries".